File: src/codes/hardcoded-variability/py/rfc_template.py
from abc import ABC, abstractmethod

class Notifier(ABC):
    def send(self, user: str, message: str):
        print(self.header())
        print(self.format_message(message))
        print(self.destination(user))

    @abstractmethod
    def header(self) -> str: pass

    @abstractmethod
    def destination(self, user: str) -> str: pass

    @abstractmethod
    def format_message(self, message: str) -> str: pass

class EmailNotifier(Notifier):
    def header(self) -> str:
        return "件名: お知らせ"

    def destination(self, user: str) -> str:
        return f"宛先: {user}@example.com"

    def format_message(self, message: str) -> str:
        return f"本文: {message}"

class SlackNotifier(Notifier):
    def header(self) -> str:
        return "Slack 通知"

    def destination(self, user: str) -> str:
        return f"Slack宛: {user}"

    def format_message(self, message: str) -> str:
        return f"内容: {message}"

# 利用例
EmailNotifier().send("hiroshi", "メールメッセージ")
SlackNotifier().send("hiroshi", "Slackメッセージ")

File: src/codes/hardcoded-variability/ts/rfc_template.ts
abstract class Notifier {
  send(user: string, message: string) {
    console.log(this.header());
    console.log(this.formatMessage(message));
    console.log(this.destination(user));
  }

  protected abstract header(): string;
  protected abstract destination(user: string): string;
  protected abstract formatMessage(message: string): string;
}

class EmailNotifier extends Notifier {
  protected header(): string {
    return "件名: お知らせ";
  }

  protected destination(user: string): string {
    return `宛先: ${user}@example.com`;
  }

  protected formatMessage(message: string): string {
    return `本文: ${message}`;
  }
}

class SlackNotifier extends Notifier {
  protected header(): string {
    return "Slack 通知";
  }

  protected destination(user: string): string {
    return `Slack宛: ${user}`;
  }

  protected formatMessage(message: string): string {
    return `内容: ${message}`;
  }
}

// 利用例
new EmailNotifier().send("hiroshi", "メールメッセージ");
new SlackNotifier().send("hiroshi", "Slackメッセージ");
